### PR TITLE
Improve release instructions

### PR DESCRIPTION
**Pull request by JelleZijlstra**
This is what I've actually been doing, but better to document it.

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
cc @AA-Turner since I think you wrote the current version of these instructions

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -30,16 +30,26 @@ backwards-incompatible changes.
 
 - Ensure that GitHub Actions reports no errors.
 
-- Update the version number in `pyproject.toml`.
+- Update the version number in `typing_extensions/pyproject.toml` and in
```
Comment (hauntsaninja) [inline CONTRIBUTING.md:33]: nit: for releases I personally find it helpful to split up steps as much as possible and not use `and`

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -30,16 +30,26 @@ backwards-incompatible changes.
...
+ - `python -m pip install --upgrade build twine`
 
 - Build the source and wheel distributions:
 
-  - `python -m pip install --upgrade build`
   - `cd typing_extensions`
   - `rm -rf dist/`
   - `python -m build .`
 
 - Install the built distributions locally and test (if you were using `tox`, you already
```
Comment (hauntsaninja) [inline CONTRIBUTING.md:48]: Might be nice to provide the commands for this

**Review by hauntsaninja — APPROVED** [review]
Looks good, feel free to merge or wait for AA-Turner :-)

File: CONTRIBUTING.md (hunk 1 above)
Comment (AA-Turner) [reply CONTRIBUTING.md:33]: I think here it is fine as the step is "update the version number", there are just two places to do it.

A

**Review by AA-Turner — APPROVED** [review]

**Comment by JelleZijlstra** [comment]
Thanks for the reviews! I'd also prefer automatic releases, since I don't exactly trust myself to do it right. (It's a lot easier with `build` than it was before, though.) On Black and a couple of other projects I maintain we generate releases automatically on tags, which works well. Could you tell me more about the option to release on PR merge? I haven't seen that before.

File: CONTRIBUTING.md (hunk 2 above)
Comment (JelleZijlstra) [reply CONTRIBUTING.md:48]: I tend to just install it and import it... I'd rather leave this out for now in the hope that we'll switch to fully automatic releases.

**Comment by AA-Turner** [comment]
> Could you tell me more about the option to release on PR merge?

The idea is entirely stolen from Hypothesis, although that project may have stolen it from somewhere else. 

The precis is that each PR includes a changelog fragment file (`RELEASE.rst` or similar), which is detected by CI and used to bump the version (major/micro/patch), prepend to the changelog, and build/tag/release.

The versioning information comes from the marker file -- either a line with release level, or release level in the suffix (`RELEASE.minor.md`).

I have an implementation for another project that I'm happy to adapt for here -- the issue is testing, really, as you don't want to be releasing thousands of versions whilst tweaking the infrastructure. Test PyPI might be a good solution for this, though.

A

**Pull request merged**